From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: granny <granny@purpurmc.org>
Date: Thu, 16 May 2024 20:37:05 -0700
Subject: [PATCH] Rebrand


diff --git a/src/main/java/io/papermc/paper/ServerBuildInfo.java b/src/main/java/io/papermc/paper/ServerBuildInfo.java
index 7196594e07af19a14c320d77df893978525fe386..a6a185d56b04df72bfd3ae55e168af7365e8d50a 100644
--- a/src/main/java/io/papermc/paper/ServerBuildInfo.java
+++ b/src/main/java/io/papermc/paper/ServerBuildInfo.java
@@ -33,6 +33,13 @@ public interface ServerBuildInfo {
     Key BRAND_PURPUR_ID = Key.key("purpurmc", "purpur");
     // Purpur end
 
+    // IceCream start
+    /**
+     * The brand id for IceCream.
+     */
+    Key BRAND_ICECREAM_ID = Key.key("icecreammc", "icecream");
+    // IceCream end
+
     /**
      * Gets the {@code ServerBuildInfo}.
      *
