From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: IceCreamMC <info@icecreammc.xyz>
Date: Wed, 07 Aug 2024 21:30:51 -7000
Subject: [PATCH] IceCream config files


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 1d594c256200d46882fafce5d2b33e60f0fa3a59..10c6f4384f33801f49aa6cfddac0b2124c0c72e7 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2266,6 +2266,13 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
         }
         // Purpur end
 
+        // IceCream start
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getIceCreamConfig() {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+        // IceCream end
+
         /**
          * Sends the component to the player
          *
