From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Pancakse <pancakse@icecreammc.xyz>
Date: Tue, 15 Oct 2024 04:33:57 +0000
Subject: [PATCH] Change Download Link For Not Updating In A While


diff --git a/src/main/java/org/bukkit/craftbukkit/Main.java b/src/main/java/org/bukkit/craftbukkit/Main.java
index d82037e8abb23f07818bc50a1f4821c10a55cab9..f3e179c5478be7457d4cd87918ed30f4a51f7bd5 100644
--- a/src/main/java/org/bukkit/craftbukkit/Main.java
+++ b/src/main/java/org/bukkit/craftbukkit/Main.java
@@ -288,8 +288,8 @@ public class Main {
                     deadline.add(Calendar.DAY_OF_YEAR, -14);
                     if (buildDate.before(deadline.getTime())) {
                         // Paper start - This is some stupid bullshit
-                        System.err.println("*** Warning, you've not updated in a while! ***");
-                        System.err.println("*** Please download a new build as per instructions from https://papermc.io/downloads/paper ***"); // Paper
+                        System.err.println("*** Warning, you've not updated IceCreamMC in a while! ***");
+                        System.err.println("*** Please download a new build of IceCream as per instructions from https://icecreammc.xyz/downloads.html ***"); // Paper // IceCream - Rebrand
                         //System.err.println("*** Server will start in 20 seconds ***");
                         //Thread.sleep(TimeUnit.SECONDS.toMillis(20));
                         // Paper end
